hyperspy/roi.py

```diff
@@ -0,0 +1,989 @@
+
+        Arguments:
+        ----------
+        signal : Signal
+            The signal to witch the widget is added. This is used to determine
+            with plot to add the widget to, and it supplies the axes_manager
+            for the widget.
+        axes : specification of axes to use, default = None
+            The axes argument specifies which axes the ROI will be applied on.
+            The DataAxis in the collection can be either of the following:
+                * "navigation" or "signal", in which the first axes of that
+                  space's axes will be used.
+                * a tuple of:
+                    - DataAxis. These will not be checked with
+                      signal.axes_manager.
+                    - anything that will index signal.axes_manager
+                * For any other value, it will check whether the navigation
+                  space can fit the right number of axis, and use that if it
+                  fits. If not, it will try the signal space.
+        widget : Widget or None (default)
+            If specified, this is the widget that will be added. If None, the
+            default widget will be used, as given by _get_widget_type().
+        color : Matplotlib color specifier (default: 'green')
+            The color for the widget. Any format that matplotlib uses should be
+            ok. This will not change the color fo any widget passed with the
+            'widget' argument.
+        """
+        axes = self._parse_axes(axes, signal.axes_manager,)
+        if widget is None:
+            widget = self._get_widget_type(axes, signal)(signal.axes_manager)
+            widget.color = color
+
+        # Remove existing ROI, if it exsists and axes match
+        if signal in self.signal_map and \
+                self.signal_map[signal][1] == axes:
+            self.remove_widget(signal)
+
+        if axes is not None:
+            # Set DataAxes
+            widget.axes = axes
+        with widget.events.changed.suppress_callback(self._on_widget_change):
+            self._apply_roi2widget(widget)
+        if widget.ax is None:
+            ax = _get_mpl_ax(signal._plot, axes)
+            widget.set_mpl_ax(ax)
+
+        # Connect widget changes to on_widget_change
+        widget.events.changed.connect(self._on_widget_change,
+                                      {'obj': 'widget'})
+        # When widget closes, remove from internal list
+        widget.events.closed.connect(self._remove_widget, {'obj': 'widget'})
+        self.widgets.add(widget)
+        self.signal_map[signal] = (widget, axes)
+        return widget
+
+    def _remove_widget(self, widget):
+        widget.events.closed.disconnect(self._remove_widget)
+        widget.events.changed.disconnect(self._on_widget_change)
+        widget.close()
+        for signal, w in self.signal_map.iteritems():
+            if w == widget:
+                self.signal_map.pop(signal)
+                break
```

Comment:
This code is quite dated, yes. Main purpose it to update `signal_map`, to remove the entry with `widget` (if present). Having one widget correspond to several entries is Bad, so it should only need to find one entry.